### PR TITLE
new post: Leveraging Rust to build cross-platform mobile libraries

### PR DESCRIPTION
**Pull request by badboy**
As it stands this is more or less a brief summary of the Glean SDK stack only.
No other major point, but I think that's good enough and in spirit with Twig

### CODE REVIEW

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+This is a rough diagram of the Glean SDK tech stack:
+
+![Glean SDK Stack](https://tmp.fnordig.de/blog/2020/glean-stack.png)
+
+On the very bottom we have `glean-core`, a pure Rust library that is the heart of the SDK.
+It's responsible for controlling the database, storing data and handling additional logic (e.g. assembling pings, clearing data, ..).
+As it is pure Rust we can rely on all Rust tooling for its development.
+We can write tests that `cargo test` picks up. We can generate the full API documentation thanks to `rustdoc`
+and we rely on `clippy` to tell us when our code is suboptimal.
+Working on `glean-core` should be possible by everyone that knows some Rust.
```
Comment (brizental) [inline _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md]:
```suggestion
Working on `glean-core` should be possible for everyone that knows some Rust.
```

Comment (brizental) [reply _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md]: This makes me wish we had a "good-first-bug"/"good-next-bug" tag to link to here.

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+On top of that sits `glean-ffi`.
+This is the FFI layer connecting `glean-core` with everything else.
+While glean-core is pure Rust, it doesn't actually provide the *nice API* we intend for users of Glean.
```
Comment (brizental) [inline _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md:63]:
```suggestion
While `glean-core` is pure Rust, it doesn't actually provide the *nice API* we intend for users of Glean.
```

Comment (brizental) [reply _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md:63]: There are some other occurances of this below as well

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+That one is later implemented on top of it all.
+glean-ffi doesn't contain much logic.
+It's a translation between the proper Rust API of glean-core and C-compatible functions exposed into the dynamic library.
+In it we rely on the excellent [`ffi-support` crate](https://docs.rs/ffi-support/).
+ffi-support knows how to translate between Rust and C types, offers a nice (and safer) abstraction for C strings.
+glean-ffi holds *some* state: the instantiated global Glean object and metric objects.
+We don't need to pass pointers back and force. Instead we use opaque handles that index into a map held inside the FFI crate.
```
Comment (brizental) [inline _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md]:
```suggestion
We don't need to pass pointers back and forth. Instead we use opaque handles that index into a map held inside the FFI crate.
```

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+The top layer of the Glean SDK are the different [language implementations](https://mozilla.github.io/glean/book/dev/core/internal/implementations.html).
+Language implementations expose a nice ergonomic API to initialize Glean and record metrics in the respective language.
+Additionally each implementation handles some special cases for the platform they are running on, like gathering application and platform data or hooking into system events.
+The nice API calls into the Glean SDK using the exposed FFI functions of `glean-ffi`.
+Unfortunately at the moment different language implementations carry different amounts of actual logic in them.
+Sometimes metric implementations require this (e.g. we rely on the clock source of Kotlin for timing metrics),
+in other parts we just didn't move the logic out of the implementations yet.
+We're actively working on moving logic into the Rust part where we can and might eventually use some code generation to unify the other parts.
```
Comment (brizental) [inline _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md]: Maybe link to [Bug 1651382](https://bugzilla.mozilla.org/show_bug.cgi?id=1651382) here?

**Review by brizental — COMMENTED** [review]
Very nice blog post, great overview really. Left some minor nits, feel free to ignore them :D

**Review by Dexterp37 — APPROVED** [review]

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md (hunk 2 above)
Comment (badboy) [reply _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md:63]: I had that before and then left it out. Inline code sticks out too much in my current design, obstructing the readability a bit.

File: _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md (hunk 3 above)
Comment (badboy) [reply _posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md]: but but ... I want to pass pointers with force!

**Pull request merged**